File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(string dartResult) {
    vector<int> sum;
    int temp = 0;
    for (int i = 0; i < dartResult.size(); i ++)
    {
        if (dartResult[i] >= '0' & dartResult[i] <= '9'){
                temp = dartResult[i]; 
                if (dartResult[i+1] == '0')
                    {
                    temp = 10;
                    i++;
                    }
            else temp = dartResult[i] - '0';
            }
        else if (dartResult[i] == 'S' || dartResult[i] == 'D' 
                 || dartResult[i] == 'T'){
            if (dartResult[i] == 'S') sum.push_back(temp);
            if (dartResult[i] == 'D') sum.push_back(temp * temp);
            if (dartResult[i] == 'T') sum.push_back(temp * temp* temp);
        }
        else if (dartResult[i] == '*'){
            if ( sum.size() == 1){
                sum[0] *=2;
            }else {
                sum[sum.size()-1] *=2;
                sum[sum.size()-2]*=2;
        }
        }else if ( dartResult[i] == '#'){
            sum[sum.size()-1] *= -1;
        }
    }
    int answer = 0;
    for ( int i = 0; i < sum.size(); i ++){
        answer += sum[i];
    }
    return answer;
}

File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    answer = 0
    score = []
    for i in range(len(dartResult)):
        if "0"<=dartResult[i]<="9": 
            result = int(dartResult[i])
            if dartResult[i+1] == "0": 
                continue
            if i-1>=0:
                if dartResult[i-1] =="1":
                    result = 10
                    
            if dartResult[i+1] == "D":
                result**=2
            elif dartResult[i+1] == "T":
                result**=3
                
            score.append(result)
            tmp=0
 
        elif dartResult[i] == "*":
            if len(score) == 1:
                score[-1]*=2
            else:
                score[-2]*=2
                score[-1]*=2
                
        elif dartResult[i] == "#":
            score[-1]*=-1     
  
            
    for i in score:
        answer+=i
        
    return answer